ENH: Resetable Funcified Methods

giovaniceotto opened the pull request:
## Pull request type

Please check the type of change your PR introduces:

- [x] Code base additions (bugfix, features)

## Pull request checklist

Please check if your PR fulfills the following requirements, depending on the type of PR:

- Code base additions (for bug fixes / features):

  - [ ] Tests for the changes have been added
  - [x] Docs have been reviewed and added / updated if needed
  - [x] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [x] All tests (`pytest --runslow`) have passed locally

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

All methods that use the `funcify_method` decorator from the Function class are cached. Once an underlying quantity that effects the value of these funcified methods is changed, they cannot be easily updated to incorporate the changes made to the class instance.

## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

A simple function called `reset_funcified_methods` has been created in the Function module. It can be imported by any multable class which makes use of the `funcify_method` decorator.

Both `HybridMotors` and `LiquidMotors` have been updated to incorporate the function `reset_funcified_methods` in end of their ´addTank´ method. 

## Does this introduce a breaking change?
<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x] No

## Other information
<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

This can be further incorporated in other parts of the code if judged as a good implementation.

This PR does not consider resetting cached properties. In a simple understanding, such methods can be decorated simply by `@property` instead of `@cached_property`. If caching is still useful, a similar approach to what was implemented in the `reset_funcified_methods`  can be done to provide the desired functionality.

phmbressan commented on rocketpy/Function.py:
```diff
@@ -3025,6 +3025,7 @@ def __get__(self, instance, owner=None):
                     )
 
                 val.__doc__ = self.__doc__
+                val.__cached__ = True
```
Great solution to mark cached attributes, I would only check if there are not any conflicts with this [standard usage](https://docs.python.org/3/reference/import.html#__cached__).

Gui-FernandesBR replied:
This link that you pointed out is related to python modules.
It is useful to show to the user where the module is located in the user's machine, specially when working with virtual environments.

Here is an example:

```bash
C:\Users\guiga>python
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
>>> numpy.__cached__
'C:\\Users\\guiga\\AppData\\Roaming\\Python\\Python310\\site-packages\\numpy\\__pycache__\\__init__.cpython-310.pyc'
>>> numpy.__file__
'C:\\Users\\guiga\\AppData\\Roaming\\Python\\Python310\\site-packages\\numpy\\__init__.py'
>>>
```

On the other hand, what Giovani has created here is a `__cached__` attribute to the funcifyed methods.
This will allow us to know if the method was cached in the class by the funcify decorator.

The `__cached__` attribute is not being assigned to any module, therefore there are not any conflicts here.

Actually, I think it is pretty intuitive using the name `__cached__` here, since it keeps a degree of consistency with the Python conventions. 

See an example:
![image](https://github.com/RocketPy-Team/RocketPy/assets/63590233/5397fb0e-3c7b-409e-b1a5-2323c14507a5)

phmbressan commented on rocketpy/Function.py:
```diff
@@ -3034,6 +3035,27 @@ def __get__(self, instance, owner=None):
+    ----------
+    instance : object
+        The instance of the class whose funcified methods will be recalculated.
+        The class must have a multable __dict__ attribute.
```
```suggestion
        The class must have a mutable __dict__ attribute.
```

phmbressan commented on rocketpy/motors/HybridMotor.py:
```diff
@@ -493,6 +493,7 @@ def addTank(self, tank, position):
         """
         self.liquid.addTank(tank, position)
         self.solid.massFlowRate = self.totalMassFlowRate - self.liquid.massFlowRate
+        reset_funcified_methods(self)
```
Really clean. Perhaps in the future we can deal with this caching in another way, but this reset really helps to reset the caching when needed and to not miss any funcify by mistake.

MateusStano commented on rocketpy/Function.py:
```diff
@@ -3034,6 +3035,27 @@ def __get__(self, instance, owner=None):
 
 
+def reset_funcified_methods(instance):
+    """Recalculates all the funcified methods of the instance. It does so by
```
```suggestion
    """Resets all the funcified methods of the instance. It does so by
```

MateusStano commented on rocketpy/Function.py:
```diff
@@ -3034,6 +3035,27 @@ def __get__(self, instance, owner=None):
         return funcify_method_decorator
 
 
+def reset_funcified_methods(instance):
+    """Recalculates all the funcified methods of the instance. It does so by
+    deleting the current Functions, which will make the interperter redefine
+    them when they are called. This is useful when the instance has changed
+    and the methods need to be recalculated.
```
This actually resets every cached_property, no? Not only funcified ones

phmbressan replied:
I don't think so. I believe the `__cached__` tag that was added only apply to `funcify`. They have no parallel in `cached_properties`.

MateusStano replied:
Okay, got it!

MateusStano commented on rocketpy/motors/HybridMotor.py:
```diff
@@ -493,6 +493,7 @@ def addTank(self, tank, position):
         """
         self.liquid.addTank(tank, position)
         self.solid.massFlowRate = self.totalMassFlowRate - self.liquid.massFlowRate
+        reset_funcified_methods(self)
```
So every time a Tank is added to a motor every cached_property of that motor object gets reset?

This probably works for the Motor classes but it can be dangerous in other places

phmbressan replied:
Could you elaborate more what do you mean by "dagerous"? My first though was exactly the opposite, all the cached parameters are being reset, even if some don't need it, therefore the code is playing safe here to ensure correctness.

MateusStano replied:
We might reset a property that does not need recalculation and is computationally heavy to calculate. Eventually, that property could need to be recalculated just because of the unnecessary reset. 

I can't think of an example where this would happen though

phmbressan replied:
I agree with you, although this is not likely to happen in methods that change fundamentally the instance, such as addTank, since most of the cache would need to be updated anyway.

I don't have a good solution for the scenario that you described. In the short term, the `del` can be used to select specific attributes to be deleted.

Gui-FernandesBR replied:
I think it is not dangerous, instead, it is inefficient.  

We don't need to be fast now, let's wait to see if any part of the code starts to increase its runtime due to this function.
We can also address it in the future.

giovaniceotto replied:
Not everytime.... Only when the methods have already been cached. Methods are only cached once called. In the general RocketPy workflow, you define everything before actually calling funcified methods.

Gui-FernandesBR commented:
For future reference, here are the classes that could start using this resetable function: 
- SolidMotor (evaluateGeometry method)
- Rocket (evaluate..... several methods)

Gui-FernandesBR reviewed (approved):
Great fix, many thanks @giovaniceotto !

I hereby confirm that I've tested in the notebooks and it was running perfectly.
Before the v1.0.0 we should definitely apply this reset feature to other classes too.

giovaniceotto commented:
Thanks for all the feedbacks.

@Gui-FernandesBR, I have added the new method to the `evaluateGeometry` method in the `SolidMotor` class. However, I have not added it to the `Rocket` class since it does not have any funcified methods.

I believe all the concerns have already been thoroughly discussed and addressed. Feel free to merge.
